lineax/_solver/Woodbury.py

```diff
@@ -0,0 +1,138 @@
+from typing import Any
+from typing_extensions import TypeAlias
+
+import jax
+import jax.numpy as jnp
+from jaxtyping import Array, PyTree
+
+from .._operator import (
+    AbstractLinearOperator,
+    is_Woodbury,
+    MatrixLinearOperator,
+    woodbury,
+)
+from .._solution import RESULTS
+from .._solve import AbstractLinearSolver, AutoLinearSolver
+from .misc import (
+    pack_structures,
+    PackedStructures,
+    ravel_vector,
+    transpose_packed_structures,
+    unravel_solution,
+)
+
+
+_Woodbury_State: TypeAlias = tuple[
+    tuple[Array, Array, Array],
+    tuple[AbstractLinearSolver, Any, AbstractLinearSolver, Any],
+    PackedStructures,
+]
+
+
+def compute_pushthrough(A_solver, A_state, C, U, V):
+    # Push through ( C^-1 + V A^-1 U) y = x
+    vmapped_solve = jax.vmap(
+        lambda x_vec: A_solver.compute(A_state, x_vec, {})[0], in_axes=1, out_axes=1
+    )
+    pushthrough_mat = jnp.linalg.inv(C) + V @ vmapped_solve(U)
+    pushthrough_op = MatrixLinearOperator(pushthrough_mat)
+    solver = AutoLinearSolver(well_posed=True).select_solver(pushthrough_op)
+    state = solver.init(pushthrough_op, {})
+    return solver, state
+
+
+class Woodbury(AbstractLinearSolver[_Woodbury_State]):
+    """Solving system using Woodbury matrix identity"""
+
+    def init(self, operator: AbstractLinearOperator, options: dict[str, Any]):
+        del options
+        if not is_Woodbury(operator):
+            raise ValueError(
+                "`Woodbury` may only be used for linear solves with A + U C V structure"
+            )
+        else:
+            A, C, U, V = woodbury(operator)
+            if A.in_size() != A.out_size():
+                raise ValueError("""A must be square""")
+            # Find correct solvers and init for A
+            A_solver = AutoLinearSolver(well_posed=True).select_solver(A)
```

Comment:
I think the choice of solver here should be an argument to `Woodbury` (the solver), defaulting to `AutoLinearSolver`.

And then at that point I don't think manually calling `select_solver` is required (or even defined in general) -- just `self.a_solver.init` should suffice.